### ISSUE TITLE
Does the test reactor need to be cached/pickled?

### ISSUE THREAD

**Issue by opotowsky** [terrapower/armi#1636]
`loadTestReactor` pickles the reactor. This caused a big bug in some downstream tests. Just wondering if this functionality is needed.

**Comment by john-science** [comment]
The reason for that is if you call `loadTestReactor()` and it has already been pickled, loading the test reactor is 20-50 times faster. It can really speed up unit tests.

What was the cause of the bug? What happened?

**Comment by john-science** [comment]
Sigh...

Arrielle has isolated the following bad situation downstream:

1. Someone builds the test reactor, and pickles it
2. The global nuclides get tweaked by some test
3. The test reactor gets unpickled
4. The test reactor is unprepared for the change in global nuclides
5. problems occur

In the long term, the solution to this issue is definitely to make the nuclides _not_ global.  We need to glue them onto the `Operator` and the `Reactor`, probably.

**Comment by ntouran** [comment]
I agree that the caching is essential to run the tests in reasonable time. However, that is a legit issue. We could cache a hash/checksum of the nuclide info and then check it upon unpack and invalidate/rebuild the cache if the hashes don't match.

**Comment by jakehader** [comment]
This might be a crazy idea, but what if we aimed to not "pickle" the reactor/operator state for testing but we targeted being able to fully define the "state" in our database so that we are very specific on what data we are holding on to. I feel like this could tend us in the direction where we have the ability to isolate interface testing for unit tests (mocking), and have more robust restart capabilities.

Saving state seems like a good idea for testing but I think we'd be able to isolate bugs/issues if we could use our common db tooling that we built to inspect the state

**Comment by opotowsky** [comment]
@ntouran, I have all the tests destroying the global nuclides after they are created (I've been rooting out these bugs over months as they appear). In this case, that did not solve the issue. 

Downstream, the only reason we haven't hit this issue before is most tests use some custom settings to load the test reactor. This means most of these tests are NOT using the cached reactor. In ARMI, there would likely be a slowdown with the removal of pickling. But I do not believe we would see a slowdown in our plugins. I'm not necessarily advocating for the removal of the pickling. I just want to understand it and see if there's a better way to control it.

**Comment by john-science** [comment]
> We could also cache a hash/checksum of the nuclide info and then check it upon unpack and invalidate/rebuild the cache if the hashes don't match.

I could implement this as a stop-gap, as soon as we are past the code freeze.

And that would buy us some time for the longer-term "global nuclides" work.

**Comment by andfranklin** [comment]
Just wanted to notify people that I'm working on a solution to this.

**Comment by john-science** [comment]
I expect this ticket to be closed when this one is:  https://github.com/terrapower/armi/issues/473

**Comment by andfranklin** [comment]
> I expect this ticket to be closed when this one is: #473

@john-science I disagree. That issue is orthogonal to this one.

I think @opotowsky's original question is great! I have a solution that makes all the spaghetti logic in `loadTestReactor` unnecessary. It will remove all the downstream bugs, maintain the caching functionality, and make that function easier to maintain.

But if you want to close this issue when you close #473 because you think you've solved all the world's problems that's fine by me 🤷‍♂️

**Comment by andfranklin** [comment]
I'd like to note that the global nuclides issue does **not** need to be solved before this one. They are orthogonal issues.

**Comment by jakehader** [comment]
@andfranklin - There was a bug due to global nuclides and pickling. If you have a solution to that fixes or can show improvements on the pickling / caching then let's do a pull request and code review. If it doesn't address the global nuclides, I'm sure that's fine, we separate out load reactor improvements from global nuclides improvements. Either way, #473 is still a problem.
